File: packages/events/python/meridian_events/outbox.py
"""Transactional outbox (SPEC 1.1): durable JSONL outbox + relay to bus."""
from __future__ import annotations

import json
import logging
import os
import threading
from pathlib import Path
from typing import Protocol

from .bus import Bus
from .envelope import Envelope

log = logging.getLogger("meridian.outbox")


class OutboxStore(Protocol):
    def append(self, topic: str, env: Envelope) -> None: ...
    def pending(self, after_seq: int, limit: int = 200) -> list[dict]: ...


class FileOutbox:
    """Durable JSONL outbox (dev default). Postgres-backed services implement
    OutboxStore with a real same-tx outbox table."""

    def __init__(self, dir_: str | os.PathLike) -> None:
        self.dir = Path(dir_)
        self.dir.mkdir(parents=True, exist_ok=True)
        self.path = self.dir / "outbox.jsonl"
        self._lock = threading.Lock()
        self._seq = 0
        if self.path.exists():
            with self.path.open() as f:
                for line in f:
                    line = line.strip()
                    if line:
                        try:
                            self._seq = max(self._seq, json.loads(line)["seq"])
                        except (json.JSONDecodeError, KeyError):
                            continue

    def append(self, topic: str, env: Envelope) -> None:
        with self._lock:
            self._seq += 1
            rec = {"seq": self._seq, "topic": topic, "envelope": env.to_dict()}
            with self.path.open("a") as f:
                f.write(json.dumps(rec) + "\n")
                f.flush()
                os.fsync(f.fileno())

    def pending(self, after_seq: int, limit: int = 200) -> list[dict]:
        out: list[dict] = []
        if not self.path.exists():
            return out
        with self._lock, self.path.open() as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    rec = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if rec.get("seq", 0) > after_seq:
                    out.append(rec)
                    if len(out) >= limit:
                        break
        return out


class OutboxRelay:
    """Drains an OutboxStore onto a Bus with a seq checkpoint."""

    def __init__(self, store: OutboxStore, bus: Bus, checkpoint_dir: str | os.PathLike,
                 interval: float = 0.5) -> None:
        self.store = store
        self.bus = bus
        self.dir = Path(checkpoint_dir)
        self.dir.mkdir(parents=True, exist_ok=True)
        self.interval = interval
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    def _checkpoint_path(self) -> Path:
        return self.dir / "outbox.relay.seq"

    def _load_checkpoint(self) -> int:
        try:
            return int(self._checkpoint_path().read_text().strip())
        except (OSError, ValueError):
            return 0

    def _save_checkpoint(self, seq: int) -> None:
        tmp = self._checkpoint_path().with_suffix(".tmp")
        tmp.write_text(str(seq))
        tmp.replace(self._checkpoint_path())

    def flush_once(self) -> int:
        seq = self._load_checkpoint()
        n = 0
        for rec in self.store.pending(seq):
            try:
                self.bus.publish(rec["topic"], Envelope.from_dict(rec["envelope"]))
            except Exception as exc:  # noqa: BLE001
                log.warning("relay publish seq=%s failed: %s", rec["seq"], exc)
                break
            seq = rec["seq"]
            n += 1
        if n:
            self._save_checkpoint(seq)
        return n

    def start(self) -> None:
        def loop() -> None:
            while not self._stop.is_set():
                self.flush_once()
                self._stop.wait(self.interval)

        self._thread = threading.Thread(target=loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=5)
        self.flush_once()

File: packages/events/outbox/outbox.go
// Package outbox implements the transactional outbox pattern (SPEC 1.1).
// In dev the canonical store is the service's embedded store and the outbox
// is a durable JSONL file; a relay goroutine drains it onto the bus.
// Services using Postgres implement Store with a real same-tx outbox table.
package outbox

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"

	"github.com/munisp/meridian-core-platform/packages/events/bus"
	"github.com/munisp/meridian-core-platform/packages/events/envelope"
)

// Record is one outbox row.
type Record struct {
	Seq      int64             `json:"seq"`
	Topic    string            `json:"topic"`
	Envelope envelope.Envelope `json:"envelope"`
}

// Store abstracts the outbox table so services can back it with Postgres.
type Store interface {
	Append(topic string, env envelope.Envelope) error
	Pending(afterSeq int64, limit int) ([]Record, error)
}

// FileStore is a durable JSONL outbox (dev default).
type FileStore struct {
	mu   sync.Mutex
	dir  string
	seq  int64
	file *os.File
}

// NewFileStore opens (or creates) a JSONL outbox in dir.
func NewFileStore(dir string) (*FileStore, error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, err
	}
	p := filepath.Join(dir, "outbox.jsonl")
	f, err := os.OpenFile(p, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return nil, err
	}
	fs := &FileStore{dir: dir, file: f}
	// recover last seq
	if rf, err := os.Open(p); err == nil {
		sc := bufio.NewScanner(rf)
		sc.Buffer(make([]byte, 1024*1024), 4*1024*1024)
		for sc.Scan() {
			var r Record
			if json.Unmarshal(sc.Bytes(), &r) == nil && r.Seq > fs.seq {
				fs.seq = r.Seq
			}
		}
		rf.Close()
	}
	return fs, nil
}

// Append writes one outbox record (durable, fsync'd).
func (fs *FileStore) Append(topic string, env envelope.Envelope) error {
	fs.mu.Lock()
	defer fs.mu.Unlock()
	fs.seq++
	line, err := json.Marshal(Record{Seq: fs.seq, Topic: topic, Envelope: env})
	if err != nil {
		return err
	}
	if _, err := fs.file.Write(append(line, '\n')); err != nil {
		return err
	}
	return fs.file.Sync()
}

// Pending returns records with seq > afterSeq.
func (fs *FileStore) Pending(afterSeq int64, limit int) ([]Record, error) {
	fs.mu.Lock()
	defer fs.mu.Unlock()
	rf, err := os.Open(filepath.Join(fs.dir, "outbox.jsonl"))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	defer rf.Close()
	var out []Record
	sc := bufio.NewScanner(rf)
	sc.Buffer(make([]byte, 1024*1024), 4*1024*1024)
	for sc.Scan() {
		var r Record
		if json.Unmarshal(sc.Bytes(), &r) == nil && r.Seq > afterSeq {
			out = append(out, r)
			if limit > 0 && len(out) >= limit {
				break
			}
		}
	}
	return out, sc.Err()
}

// Close flushes and closes the outbox file.
func (fs *FileStore) Close() error {
	fs.mu.Lock()
	defer fs.mu.Unlock()
	return fs.file.Close()
}

// Relay drains an outbox Store onto a Bus, checkpointing progress.
type Relay struct {
	Store    Store
	Bus      bus.Bus
	Dir      string // checkpoint file location
	Interval time.Duration
}

// Run starts the relay loop until ctx is cancelled. Non-blocking errors are logged.
func (r *Relay) Run(ctx context.Context) {
	interval := r.Interval
	if interval <= 0 {
		interval = 500 * time.Millisecond
	}
	t := time.NewTicker(interval)
	defer t.Stop()
	for {
		r.flushOnce(ctx)
		select {
		case <-ctx.Done():
			r.flushOnce(ctx) // final drain
			return
		case <-t.C:
		}
	}
}

func (r *Relay) checkpointPath() string { return filepath.Join(r.Dir, "outbox.relay.seq") }

func (r *Relay) loadCheckpoint() int64 {
	b, err := os.ReadFile(r.checkpointPath())
	if err != nil {
		return 0
	}
	var n int64
	fmt.Sscanf(string(b), "%d", &n)
	return n
}

func (r *Relay) saveCheckpoint(n int64) {
	tmp := r.checkpointPath() + ".tmp"
	if err := os.WriteFile(tmp, []byte(fmt.Sprintf("%d", n)), 0o644); err == nil {
		os.Rename(tmp, r.checkpointPath())
	}
}

func (r *Relay) flushOnce(ctx context.Context) {
	seq := r.loadCheckpoint()
	recs, err := r.Store.Pending(seq, 200)
	if err != nil {
		log.Printf("outbox relay: read pending: %v", err)
		return
	}
	for _, rec := range recs {
		if err := r.Bus.Publish(ctx, rec.Topic, rec.Envelope); err != nil {
			log.Printf("outbox relay: publish seq=%d topic=%s: %v", rec.Seq, rec.Topic, err)
			return // retry next tick
		}
		seq = rec.Seq
	}
	if len(recs) > 0 {
		r.saveCheckpoint(seq)
	}
}
